File: backend/api/v1/wishlist.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlmodel import Session, select
from typing import List
from db import get_session
from models.user import User
from models.product import Product
from models.wishlist import Wishlist
from api.deps import get_current_active_user

router = APIRouter()

@router.get("/", response_model=List[int])
def get_wishlist(
    *, session: Session = Depends(get_session),
    current_user: User = Depends(get_current_active_user)
):
    """
    Retorna los IDs de los productos en la lista de deseos del usuario.
    """
    statement = select(Wishlist.product_id).where(Wishlist.user_id == current_user.id)
    items = session.exec(statement).all()
    return items

@router.post("/toggle/{product_id}")
def toggle_wishlist_item(
    *, session: Session = Depends(get_session),
    product_id: int,
    current_user: User = Depends(get_current_active_user)
):
    """
    Agrega o elimina un producto de la lista de deseos.
    """
    # Verificar que el producto exista
    product = session.get(Product, product_id)
    if not product:
        raise HTTPException(status_code=404, detail="Producto no encontrado")
        
    statement = select(Wishlist).where(Wishlist.user_id == current_user.id, Wishlist.product_id == product_id)
    existing_item = session.exec(statement).first()
    
    if existing_item:
        session.delete(existing_item)
        session.commit()
        return {"action": "removed", "productId": product_id}
    else:
        new_item = Wishlist(user_id=current_user.id, product_id=product_id)
        session.add(new_item)
        session.commit()
        return {"action": "added", "productId": product_id}

@router.get("/check/{product_id}")
def check_wishlist_item(
    *, session: Session = Depends(get_session),
    product_id: int,
    current_user: User = Depends(get_current_active_user)
):
    """
    Verifica si un producto está en la lista de deseos.
    """
    statement = select(Wishlist).where(Wishlist.user_id == current_user.id, Wishlist.product_id == product_id)
    item = session.exec(statement).first()
    return {"inWishlist": item is not None}

File: backend2/src/routes/wishlist.ts
/**
 * Rutas de Favoritos (Wishlist) — /api/v1/wishlist
 *
 * Endpoints:
 *   GET    /                — Obtener IDs de productos en la wishlist del usuario
 *   POST   /toggle/:id      — Agregar/Quitar un producto de la wishlist
 */

import { Hono } from "hono";
import { eq, and } from "drizzle-orm";
import { createDb } from "../db/index";
import { wishlistItems, products } from "../db/schema";
import { requireAuth } from "../middleware/auth";
import type { Env, AppVariables } from "../types";

const wishlistRouter = new Hono<{
  Bindings: Env;
  Variables: AppVariables;
}>();

// Todas las rutas requieren autenticación
wishlistRouter.use("/*", requireAuth);

// ─── GET / ──────────────────────────────────────────────

/**
 * Obtener todos los product_id en la wishlist del usuario autenticado.
 * Retorna un arreglo de IDs: [1, 2, 5]
 */
wishlistRouter.get("/", async (c) => {
  const currentUser = c.get("currentUser")!;
  const db = createDb(c.env.DB);

  const items = await db.query.wishlistItems.findMany({
    where: eq(wishlistItems.user_id, currentUser.id),
    columns: {
      product_id: true,
    },
  });

  const ids = items.map((item) => item.product_id);
  return c.json(ids);
});

// ─── POST /toggle/:id ───────────────────────────────────

/**
 * Agrega o elimina un producto de la wishlist del usuario.
 * Si ya existe, lo quita. Si no existe, lo agrega.
 */
wishlistRouter.post("/toggle/:id", async (c) => {
  const productId = Number(c.req.param("id"));
  const currentUser = c.get("currentUser")!;
  const db = createDb(c.env.DB);

  // Verificar que el producto exista
  const product = await db.query.products.findFirst({
    where: eq(products.id, productId),
  });

  if (!product) {
    return c.json({ detail: "Producto no encontrado" }, 404);
  }

  // Verificar si ya está en la wishlist
  const existing = await db.query.wishlistItems.findFirst({
    where: and(
      eq(wishlistItems.user_id, currentUser.id),
      eq(wishlistItems.product_id, productId)
    ),
  });

  if (existing) {
    // Eliminarlo
    await db.delete(wishlistItems).where(eq(wishlistItems.id, existing.id));
    return c.json({ status: "removed", product_id: productId });
  } else {
    // Agregarlo
    await db.insert(wishlistItems).values({
      user_id: currentUser.id,
      product_id: productId,
    });
    return c.json({ status: "added", product_id: productId }, 201);
  }
});

export default wishlistRouter;
